dimod/binary_quadratic_model.py

```diff
@@ -236,32 +246,40 @@ def __repr__(self):
 
     def __eq__(self, other):
         """Model is equal if and only if linear, adj, offset and vartype are all equal."""
-        if not isinstance(other, BinaryQuadraticModel):
-            return False
 
-        # NB: adj is invariant of edge order, so check that instead of quadratic
-        return (self.vartype == other.vartype and self.offset == other.offset
-                and self.linear == other.linear and self.adj == other.adj)
+        try:
+            if self.vartype is not other.vartype:
+                return False
+
+            if self.offset != other.offset:
+                return False
+
+            if self.linear != other.linear:
+                return False
+
+            return self.adj == other.adj
+        except AttributeError:
+            return False
 
     def __ne__(self, other):
-        """Inversion of equality."""
-        return not self.__eq__(other)
+        return not (self == other)
 
     def __len__(self):
-        """The length is number of variables."""
-        return self.adj.__len__()
+        return len(self.adj)
 
     def __contains__(self, v):
-        """The variables"""
-        return self.adj.__contains__(v)
+        return v in self.adj
 
     def __iter__(self):
-        return self.adj.__iter__()
+        return iter(self.adj)
 
     @property
     def variables(self):
         """Return binary quadratic model's variables as a dictionary view object."""
-        return viewkeys(self.linear)
+        if PY2:
+            return set(self.linear)
+        else:
+            return self.linear.keys()
```

Comment:
Since `LinearView` is a mapping, we could add `.keys()` there (as a proxy to `adj.keys()`), and then here use:
```
return self.linear
```
Would that work?

Reply:
That would work, things like
`1 in bqm.variables` would work fine, but if you did `print(bqm.variables)` it might be confusing.